### PR TITLE
Implement attribute getter of Location #5591 

### PR DESCRIPTION
**Pull request by shinglyu**
Reopening @yodalee 's https://github.com/servo/servo/pull/5597

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6135)

### CODE REVIEW

File: components/script/dom/urlhelper.rs [hunk 1]
```diff
@@ -34,6 +34,59 @@ impl UrlHelper {
...
+            Some(password) => password.to_owned()
+        })
+    }
+
+    pub fn Host(url: &Url) -> USVString {
+        // https://url.spec.whatwg.org/#dom-urlutils-host
+        // FIXME: Url null check is skipped for now
+        let host = match url.host() {
+            None => "".to_owned(),
+            Some(host) => format!("{}", host)
```
Comment (Manishearth) [inline components/script/dom/urlhelper.rs]: nit: we format `host` twice here, perhaps we shouldn't (instead do all formatting in the match below)

**Comment by Manishearth** [comment]
Could you file a followup about making these attributes writeable, and about fixing the nullable-URL thing? (the current implementation using `Window` is wrong; I think.

**Comment by shinglyu** [comment]
Per @Ms2ger's suggestion, I'll add a test to data url. 

I think I'll open a followup for writable and nullable url. I don't want to block other people for too long. :)

**Comment by shinglyu** [comment]
@Ms2ger : It turns out that there are two problems the prevent the data url test for location getters:

1. #5500 prevents me from using `iframe.contentWindow.location`, the `contentWindow` will be `null` if I don't use a timeout
2. Most of the location getters will return `""`, but the null check is not enabled. So I'll enable them (and add the test) in the followup bug.

**Comment by shinglyu** [comment]
Review status: 0 of 11 files reviewed, 1 unresolved discussion, all commit checks successful.

---

<sup>**[components/script/dom/urlhelper.rs, line 65 \[r2\]](https://reviewable.io:443/reviews/servo/servo/6135#-Jpgx9l6zE_su1ZAPCpQ-r2-65)** ([raw file](https://github.com/servo/servo/blob/23e99966e43753a6af196b3d0661207aa8a0438d/components/script/dom/urlhelper.rs#L65)):</sup>
Done. (I'm not sure about how to use Reviewable, should I mark this as done manually?)

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6135)

**Comment by jdm** [comment]
@shinglyu You can click the "acknowledge" button to mark issues as resolved in Reviewable.

**Comment by Manishearth** [comment]
Review status: :shipit: all files reviewed, all discussions resolved, all commit checks successful.
Reviewed files:
- components/script/dom/location.rs @ r1
- components/script/dom/urlhelper.rs @ r3
- components/script/dom/webidls/URLUtils.webidl @ r1
- tests/wpt/metadata/html/browsers/history/the-location-interface/location_host.html.ini @ r2
- tests/wpt/metadata/html/browsers/history/the-location-interface/location_hostname.html.ini @ r2
- tests/wpt/metadata/html/browsers/history/the-location-interface/location_pathname.html.ini @ r2
- tests/wpt/metadata/html/browsers/history/the-location-interface/location_port.html.ini @ r2
- tests/wpt/metadata/html/browsers/history/the-location-interface/location_protocol.html.ini @ r2
- tests/wpt/metadata/XMLHttpRequest/send-authentication-basic-cors.htm.ini @ r2
- tests/wpt/metadata/XMLHttpRequest/send-authentication-basic-setrequestheader.htm.ini @ r2
- tests/wpt/metadata/XMLHttpRequest/timeout-cors-async.htm.ini @ r2

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6135)

**Comment by Manishearth** [comment]
```

/Users/servo/buildbot/slave/mac1/build/components/script/dom/urlhelper.rs:65:13: 65:17 error: expected one of `,`, `.`, `}`, or an operator, found `Some`
/Users/servo/buildbot/slave/mac1/build/components/script/dom/urlhelper.rs:65             Some(host) => host
```

**Comment by shinglyu** [comment]
Oops, let me fix that right away

**Comment by jdm** [comment]
@shinglyu Could you squash the compile error into the original commit?

**Comment by shinglyu** [comment]
Sure, I'll do that when I get home.

Josh Matthews <notifications@github.com> 於 2015年6月2日 週二 上午5:57寫道：

**Comment by shinglyu** [comment]
I remeber I dealt with those tests before, let me check. Sorry!

**Comment by jdm** [comment]
```
/XMLHttpRequest/send-after-setting-document-domain.htm
------------------------------------------------------
FAIL loading documents from original origin after setting document.domain
OK expected TIMEOUT [Parent]
/XMLHttpRequest/timeout-cors-async.htm
--------------------------------------
FAIL XMLHttpRequest: timeout event and cross-origin request
/workers/WorkerLocation_host.htm
--------------------------------
FAIL  WorkerLocation URL decomposition IDL attribute: host 
/workers/WorkerLocation_hostname.htm
------------------------------------
FAIL  WorkerLocation URL decomposition IDL attribute: hostname 
/workers/WorkerLocation_port.htm
--------------------------------
FAIL  WorkerLocation URL decomposition IDL attribute: port 
/workers/WorkerLocation_protocol.htm
------------------------------------
FAIL  WorkerLocation URL decomposition IDL attribute: protocol 
```

**Comment by shinglyu** [comment]
I disabled those tests locally, but didn't commit the `.ini`s. Let me check if they should be disabled or not.

**Comment by jdm** [comment]
Disabling (ie. not running) is the wrong choice here. Those are valid results; the tests now check more things because they were previously stopping silently on properties of location that didn't exist.

**Comment by snf** [comment]
These tests are failing because WorkerLocation implement URLUtilsReadOnly which this patch does not include. These was passing before because both (URLUtils and URLUtilsReadOnly) returned `undefined`. I think the right choice would be to make the tests fail and wait for another patch to implement URLUtilsReadOnly.
I was working in this but I noticed this PR so I will work it from here (also have other patches for Worker)

**Comment by shinglyu** [comment]
Fernandez, thanks for the information, I'll disable them.

Sebastian N. Fernandez <notifications@github.com> 於 2015年6月4日 週四 上午12:54寫道：

**Comment by snf** [comment]
@shinglyu I think you have to change them to FAIL, not disable them.

**Comment by shinglyu** [comment]
@snf: sorry, I mean turn them to expected FAIL.

**Comment by shinglyu** [comment]
Seems like the WorkerLocation tests are failing due to the reason mentioned by @snf .
The XMLHttpRequest is failing because `document.domain` is `undefined`, which seems to be caused by https://github.com/servo/servo/issues/934

So can I change them to expected `FAIL` for now?

**Comment by Ms2ger** [comment]
Sure

**Comment by nox** [comment]
Err, didn't realise that my #6371 supersedes this.

**Comment by shinglyu** [comment]
@nox: sorry, I should have make this merge earlier.

**Comment by vectorijk** [comment]
@shinglyu I am a newbie and diving into how to implement the history api(forward and back). It may be helpful for me to add origin( https://url.spec.whatwg.org/#dom-urlutils-origin) into urlhelper.rs. Is that reasonable? Thanks~

**Comment by shinglyu** [comment]
@vectorijk : I am also quite new to this area, I think is reasonable, but you might need to consult someone who is more experienced then me.

**Comment by shinglyu** [comment]
#6371 implemented this. Should I close it?

**Comment by jdm** [comment]
I guess so!

**Pull request closed without merging**